File: 86-partitionList/partition.kt
class ListNode(var `val`: Int) {
    var next: ListNode? = null
}

class Solution {
    fun partition(head: ListNode?, x: Int): ListNode? {
        val s = ListNode()
        var s_curr = s
        val be = ListNode()
        var be_curr = be
        head.forEach {node -> if (node.`val` < x) {
                    s_curr.next = node 
                    s_curr = s_curr.next
                } else {
                    be_curr.next = node
                    be_curr = be_curr.next
                }
            }
        be_curr.next = null
        s_curr.next = be.next
        return s.next
    }
    fun ListNode?.forEach(f: (ListNode) -> Unit) {
        if (this == null) {
            return
        }
        f(this)
        this.next.forEach(f)
    }
}

File: 86-partitionList/partition.go
package partition

type ListNode struct {
	Val  int
	Next *ListNode
}

func partition(head *ListNode, x int) *ListNode {
	s := &ListNode{}
	be := &ListNode{}
	s_curr := s
	be_curr := be

	for head != nil {
		if head.Val < x { // attach to s list
			s_curr.Next = head
			s_curr = s_curr.Next
		} else { // attach to bigger-equal list
			be_curr.Next = head
			be_curr = be_curr.Next
		}
		head = head.Next
	}

	be_curr.Next = nil    // avoid loops
	s_curr.Next = be.Next // concat smaller with bigger

	return s.Next
}
